## Homework 6 Question 1
##### 2022-07-10

In [69]:
import time
import numpy as np
import pandas as pd
import scipy.sparse as sp

### Part A

In [83]:
df = pd.read_csv("TexasCityDistanceMatrix.csv", index_col = 0, encoding = "latin-1")
key = pd.read_csv("TexasKey.csv", index_col = 0, encoding = "latin-1")

print(df.max().max())

99.99971353


The maximum value in the matrix is 99.9997

This implies that the threshold cut-off for distance is 100mi. Therefore, the missing values in the matrix are the distances between two locations that are greater than 100.

### Part B

In [71]:
katy_id = key.loc[key['Name'] == 'Katy city'].index[0]
closest = sorted(df.loc[katy_id,:][df.loc[katy_id,:].notna()])[1]
closest_id_string = df.loc[katy_id,:][np.where(df.loc[katy_id,:] == closest)[0]].index[0]
closest_city = key.loc[int(closest_id_string[1:]), 'Name']
print(closest_city)

Cinco Ranch CDP


In [72]:
flower_id = key.loc[key['Name'] == 'Flower Mound town'].index[0]
print(len(df.loc[flower_id,:][df.loc[flower_id,:] < 25]))

76


The closest location to "Katy city" is "Cinco Ranch CDP" and there are 76 locations within 25 miles of "Flower Mound town."

### Part C

In [78]:
start = time.perf_counter()
np.nansum(df)
print(time.perf_counter() - start)

0.02037630000268109


It takes roughly 20 milliseconds to sum the entire matrix.

### Part D

In [89]:
df[df > 50] = 0
df.fillna(0, inplace = True)

start = time.perf_counter()
np.nansum(df)
print(time.perf_counter() - start)

0.00816359999589622


It takes roughtly 8 milliseconds to sum the entire new matrix. This is greater than a 50% reduction in computation time.

### Part E

In [108]:
df = pd.read_csv("TexasCityDistanceMatrix.csv", index_col = 0, encoding = "latin-1")

# using -1 as placeholder to swap 0s with NaNs
df[df == 0] = -1
df[np.isnan(df)] = 0
df[df == -1] = np.nan

df = sp.csr_matrix(df)

# putting 0s back
df.data[np.isnan(df.data)] = 0

start = time.perf_counter()
df.sum()
print(time.perf_counter() - start)

0.0007184999994933605


It takes roughtly 0.7 milliseconds to sum the entire sparse matrix. This is greater than a 90% reduction in computation time from the cleaned matrix and an incredible 96% reduction from the initial matrix.

### Part F

Given that the zeroes are distances of cities to themselves, they are unnecessary and can be removed. Not only this, the distances in the matrix are symmetric; there is no need to track both the distance from point A to point B and the distance from point B to point A as they will always be the same. Removing these duplicates would cut the memory consumption in half.